tools/data-handler/src/utils/configuration-logger.ts

```diff
@@ -0,0 +1,267 @@
+  You should have received a copy of the GNU Affero General Public
+  License along with this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import { readFile, rename } from 'node:fs/promises';
+import { join } from 'node:path';
+
+import type { Logger } from 'pino';
+
+import { getChildLogger } from './log-utils.js';
+import { ProjectPaths } from '../containers/project/project-paths.js';
+import { writeFileSafe, pathExists } from './file-utils.js';
+
+/**
+ * Enum for configuration change operation types.
+ */
+export enum ConfigurationOperation {
+  MODULE_ADD = 'module_add',
+  MODULE_REMOVE = 'module_remove',
+  PROJECT_RENAME = 'project_rename',
+  RESOURCE_CREATE = 'resource_create',
+  RESOURCE_DELETE = 'resource_delete',
+  RESOURCE_RENAME = 'resource_rename',
+  RESOURCE_UPDATE = 'resource_update',
+}
+
+/**
+ * Individual log entry representing a single configuration change.
+ * @param timestamp Timestamp when the operation occurred (ISO string)
+ * @param operation Type of operation performed
+ * @param target Target of the operation
+ * @param parameters Additional parameters specific to the operation
+ */
+export interface ConfigurationLogEntry {
+  timestamp: string;
+  operation: ConfigurationOperation;
+  target: string;
+  parameters?: Record<string, unknown>;
+}
+
+/**
+ * Options for logging configuration changes.
+ * @param parameters  Additional parameters to store with the operation
+ */
+export interface ConfigurationLogOptions {
+  parameters?: Record<string, unknown>;
+}
+
+/**
+ * Logger for tracking configuration changes that affect project structure.
+ */
+export class ConfigurationLogger {
+  private static instance: ConfigurationLogger | null = null;
+  private readonly paths: ProjectPaths;
+  private readonly logger: Logger;
+  private readonly logFile: string;
+
+  /**
+   * Constructor - Creates an instance of ConfigurationLogger
+   * @param projectPath Path to a project root
+   * @param logger Logger instance; creates new if missing
+   */
+  private constructor(
+    private projectPath: string,
+    logger?: Logger,
+  ) {
+    this.logger = logger || getChildLogger({ module: 'ConfigurationLogger' });
+    this.paths = new ProjectPaths(projectPath);
+    this.logFile = this.paths.configurationChangesLog;
+  }
+
+  // Internal async method that does the actual logging.
+  private async doLog(
+    operation: ConfigurationOperation,
+    target: string,
+    options?: ConfigurationLogOptions,
+  ): Promise<void> {
+    try {
+      const entry: ConfigurationLogEntry = {
+        timestamp: new Date().toISOString(),
+        operation,
+        target,
+        parameters: options?.parameters,
+      };
+
+      await writeFileSafe(this.logFile, JSON.stringify(entry) + '\n', {
+        flag: 'a',
+      });
+
+      this.logger.debug(`Logged ${operation} operation for target: ${target}`);
+    } catch (error) {
+      this.logger.error({ error }, `Failed to log change`);
```

Comment:
In addition to error, could add operation and target to the structured log

```suggestion
      this.logger.error({ error, target, operation }, `Failed to log change`);
```
Logs can then be filtered by them